File: 4.lua
io.input("4.txt")

function show_board(n, board, selection)
    print()
    print(string.format("%d)", n), 1, 2, 3, 4, 5)
    print(" ", "--", "--", "--", "--", "--")
    for r, row in ipairs(board) do
        local numbers = {}
        for c, number in ipairs(row) do
            numbers[c] = selection[r][c] and "*" .. number or number
        end
        print(r .. "|", table.unpack(numbers))
    end
end

function copy_grid(grid)
    local copy = {}

    for i = 1, 5 do
        copy[i] = {}
        for j = 1, 5 do
            copy[i][j] = grid[i][j]
        end
    end

    return copy
end

function pos_of(number, board)
    for i, r in pairs(board) do
        for j, n in ipairs(r) do
            if n == number then
                return i, j
            end
        end
    end
end

function check_winner(selections, winners)
    local new_winners = {}

    for b, _ in ipairs(selections) do
        local has_won
        for _, winner in ipairs(winners) do
            if winner["n"] == b then
                has_won = true
                break
            end
        end

        if not has_won then
            local rows = { 0, 0, 0, 0, 0 }
            local cols = { 0, 0, 0, 0, 0 }
            for r = 1, 5 do
                for c = 1, 5 do
                    if selections[b][r][c] then
                        rows[r] = rows[r] + 1
                        cols[c] = cols[c] + 1
                        if cols[c] == 5 or rows[r] == 5 then
                            new_winners[#new_winners + 1] = b
                        end
                    end
                end
            end
        end
    end

    return new_winners
end

function calculate_score(board, selection, pick)
    local score = 0
    for r, row in ipairs(board) do
        for c, number in ipairs(row) do
            if not selection[r][c] then
                score = score + number
            end
        end
    end

    return score * pick
end

local picks = {}
for pick in string.gmatch(io.read("l"), "(%d+),?") do
    picks[#picks + 1] = tonumber(pick)
end

local boards = {}
for line in io.lines() do
    if not line then
        break
    end

    if line == "" then
        boards[#boards + 1] = {}
    else
        local row = {}
        for n in string.gmatch(line, "(%d+)") do
            row[#row + 1] = tonumber(n)
        end
        boards[#boards][#boards[#boards] + 1] = row
    end
end

local selections = {}
for b, _ in ipairs(boards) do
    selections[b] = {}
    for r = 1, 5 do
        selections[b][r] = {}
    end
end

local winners = {}
for i, pick in ipairs(picks) do
    for b, board in ipairs(boards) do
        local r, c
        r, c = pos_of(pick, board)
        if r and c then
            selections[b][r][c] = true
        end
    end

    local new_winners = check_winner(selections, winners)

    for _, winner in ipairs(new_winners) do
        winners[#winners + 1] = {
            n = winner,
            score = calculate_score(boards[winner], selections[winner], pick),
            pick = pick,
            selection = copy_grid(selections[winner]),
        }
    end
end

print(string.format("We have %d winners out of %d boards", #winners, #boards))

print(
    string.format(
        "\n\nThe first winner is board %d with a score of %d and the winning pick of %d.",
        winners[1]["n"],
        winners[1]["score"],
        winners[1]["pick"]
    )
)
show_board(winners[1]['n'], boards[winners[1]['n']], winners[1]['selection'])

print(
    string.format(
        "\n\nThe last winner is board %d with a score of %d and the winning pick of %d.",
        winners[#winners]["n"],
        winners[#winners]["score"],
        winners[#winners]["pick"]
    )
)
show_board(winners[#winners]['n'], boards[winners[#winners]['n']], winners[#winners]['selection'])

File: 4.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INPUT "4.txt"
#define N_ROWS 5
#define N_COLS 5
#define LEN_BOARD 25
#define MAX_BOARDS 100
#define MAX_NUMBERS 100

typedef struct {
	int b, n;
} Winner;

static bool check_board(int *board) {
	int rows[N_ROWS] = {0};
	int cols[N_COLS] = {0};

	for (int row = 0; row < N_ROWS; row++) {
		for (int col = 0; col < N_COLS; col++)
			if (board[row * 5 + col] == -1) {
				rows[row]++;
				cols[col]++;
				if (rows[row] == 5 || cols[col] == 5)
					return true;
			}
	}
	return false;
}

bool mark_boards(int *boards, int len, int number, Winner *ret) {
	int board = 0;

	for (int i = 0; i < len; i++) {
		if (boards[i] == number) {
			board = i / LEN_BOARD;
			boards[i] = -1;
			if (check_board(boards + LEN_BOARD * board)) {
				*ret = (Winner){.b = board, .n = number};
				return true;
			}
		}
	}

	return false;
}

static int calculate_score(const int *board, int number) {
	int sum_unmarked = 0;
	for (int i = 0; i < LEN_BOARD; i++) {
		if (board[i] != -1)
			sum_unmarked += board[i];
	}

	return sum_unmarked * number;
}

void one(const int *numbers, int n_numbers, const int *boards, int n_boards) {
	printf("--- Part One ---\n");
	int *boards_cpy = malloc(sizeof(int) * n_boards);
	memcpy(boards_cpy, boards, sizeof(int) * n_boards);

	Winner winner;
	for (int n = 0; n < n_numbers; n++) {
		for (int i = 0; i < n_boards; i++) {
			if (boards_cpy[i] == numbers[n]) {
				boards_cpy[i] = -1;
				if (check_board(boards_cpy +
				                (i / LEN_BOARD) * LEN_BOARD)) {
					winner = (Winner){.b = i / LEN_BOARD,
					                  .n = numbers[n]};
					goto finally;
				}
			}
		}
	}

finally:
	printf("1st: %2d / %2d => %d\n", winner.b, winner.n,
	       calculate_score(boards_cpy + winner.b * LEN_BOARD, winner.n));

	free(boards_cpy);
}

void two(const int *numbers, int n_numbers, const int *boards, int n_boards) {
	printf("--- Part Two ---\n");
	int *boards_cpy = malloc(sizeof(int) * n_boards);
	memcpy(boards_cpy, boards, sizeof(int) * n_boards);

	int n_winners = 0;
	Winner winners[MAX_BOARDS] = {0};

	for (int n = 0; n < n_numbers; n++) {
		for (int i = 0; i < n_boards; i++) {
			for (int w = 0; w < n_winners; w++)
				if (winners[w].b == i / LEN_BOARD)
					goto next;
			if (boards_cpy[i] == numbers[n]) {
				boards_cpy[i] = -1;
				if (check_board(boards_cpy +
				                (i / LEN_BOARD) * LEN_BOARD)) {
					winners[n_winners++] =
					    (Winner){.b = i / LEN_BOARD,
					             .n = numbers[n]};
					continue;
				}
			}
		next:;
		}
	}

	Winner first = winners[0];
	Winner last = winners[n_winners - 1];

	printf("1st: %2d / %2d => %d\n", first.b, first.n,
	       calculate_score(boards_cpy + first.b * LEN_BOARD, first.n));
	printf("nth: %2d / %2d => %d\n", last.b, last.n,
	       calculate_score(boards_cpy + last.b * LEN_BOARD, last.n));

	free(boards_cpy);
}

int main() {
	// Read the input file into a dynamically allocated string buffer.
	FILE *file = fopen(INPUT, "r");
	fseek(file, 0L, SEEK_END);
	size_t size = (size_t)ftell(file);
	fseek(file, 0L, SEEK_SET);
	char *content = malloc(size + 1);
	char *tofree = content;
	size_t bytes_read = fread(content, sizeof(char), size, file);
	content[bytes_read] = '\0';
	fclose(file);

	// Setup the input data
	int n_numbers = 0;
	int numbers[MAX_NUMBERS];
	char *number_line = strsep(&content, "\n");
	for (char *num; (num = strsep(&number_line, ", "));)
		numbers[n_numbers++] = atoi(num);

	int n_boards = 0;
	int boards[LEN_BOARD * MAX_BOARDS] = {0};
	for (char *num; (num = strsep(&content, " \n"));)
		if (strcmp(num, "") != 0)
			boards[n_boards++] = atoi(num);

	one(numbers, n_numbers, boards, n_boards);
	two(numbers, n_numbers, boards, n_boards);

	free(tofree);
	return 0;
}
